src/generate_opencl_dynamic_loader.py

```diff
@@ -0,0 +1,176 @@
+            print(f"Function '{function_name}' skipped give api={x[4]}")
+            continue
+        
+        # Begin function
+        dynamic_loader.write(f'/* {function_name} */\nstatic {function_return} (*ptr_{function_name})({function_params}) = NULL;\n')    # Function pointer definition
+        dynamic_loader.write(f'CL_API_ENTRY {function_return} CL_API_CALL {function_name}({function_params})\n') # Function definition
+        dynamic_loader.write('{\n\t')
+        
+        # If we are in the special case
+        if function_name == special_function_name:
+            dynamic_loader.write(special_function_code)
+        dynamic_loader.write(f'return ptr_{function_name}(') # Function call through pointer
+        
+        # Manage params
+        function_params = "".join(re.split('\(CL_CALLBACK *', function_params))
+        function_params = "".join(re.split('\)\([_a-zA-Z0-9 \*,]+\)', function_params))
+        param_names = re.findall(f'(const )?(unsigned )?{REGEX_C_ID}[ \*]+({REGEX_C_ID})', function_params)
+        for i in range(len(param_names)):
+            dynamic_loader.write(f'{", " if i > 0 else ""}{param_names[i][2]}')
+        
+        # End function
+        dynamic_loader.write(');\n}\n\n')
+    else:
+        print("Error parsing CL.h header file")
+        exit(1)
+        
+# Load dynamic library
+dynamic_loader.write(
+'''
+#include <dlfcn.h>
+#include <stdio.h>
+static void load_opencl_dll()
+{
+        int i;
+        if (opencl_dll)
+            return;
+
+        // Names to try to load
+        const char* opencl_names[] = {
+            "libOpenCL.so",      // Linux/others
+            "OpenCL",            // _WIN
+            "/System/Library/Frameworks/OpenCL.framework/OpenCL", // __APPLE__
+            "opencl.dll",        // __CYGWIN__
+            "cygOpenCL-1.dll",   // __CYGWIN__
+            "libOpenCL.so.1"     // Linux/others
+        };
+        for (i = 0; i < sizeof(opencl_names)/sizeof(opencl_names[0]); i++)
+        {
+            opencl_dll = dlopen(opencl_names[i], RTLD_NOW);
+            if (opencl_dll) break;
+        }      
+          
+        // Load function pointers
+        if (opencl_dll)
+        {
+                int all_functions_loaded = 1;
+
+''')
+
+# Load function pointers
+for x in funtions:
+    function_name = x[2]
+    api_version = int(x[4].replace('_', '')) * 10
+    if api_version <= CL_TARGET_OPENCL_VERSION:
+        dynamic_loader.write(f'\t\tptr_{function_name} = dlsym(opencl_dll, "{function_name}");\n')
+        dynamic_loader.write(f'\t\tif (!ptr_{function_name})\n')
+        dynamic_loader.write('\t\t{\n')
+        dynamic_loader.write(f'\t\t\tall_functions_loaded = 0;\n')
+        dynamic_loader.write(f'\t\t\tprintf("Cannot load {function_name} function\\n");\n')
+        dynamic_loader.write('\t\t}\n')
+
+dynamic_loader.write('''
+            if (!all_functions_loaded)
+            {
+                dlclose(opencl_dll);

```

Comment:
Here we insist on finding all functions for our target OpenCL version. I suppose someone could want to get JtR working on a partial OpenCL implementation, which does not yet have all the functions, but maybe happens to have all those that we actually use. We could want to re-add a way to do that - somehow lookup/require only functions that we actually use or/and re-add a way (e.g., a `configure` option) to link against OpenCL like we did before.

Reply:
We can add a check on the Python script generator to only generate code/checks for functions we use, but all contributors need to know if they will use in the future some non-common function, JtR will not compile. Do you want that I make these changes?

Reply:
> We can add a check on the Python script generator to only generate code/checks for functions we use, but all contributors need to know if they will use in the future some non-common OpenCL function, JtR will not compile. Do you want that I make these changes?

I went with somewhat different approach in #5395, please help test and review it. Thank you!